pkg/storage/db.go

```diff
@@ -199,6 +208,79 @@ func (db *DB) ping() error {
 	return nil
 }
 
+// GeneratePartition finds the latest partition, and generate the next partition based on the previous partions time range
+func (db *DB) GeneratePartition(ctx context.Context) error {
+	stmt := "SELECT partition_end FROM partitions ORDER BY partition_begin DESC LIMIT 1"
+	row := db.sqldb.QueryRowContext(ctx, stmt)
+	var latest string
+	err := row.Scan(&latest)
+	switch err {
+	case nil:
+	case sql.ErrNoRows:
+		return db.GeneratePartitionWithTimestamp(ctx, db.now())
+	default:
+		return err
+	}
+
+	latestTime, err := time.Parse(time.RFC3339, latest)
+	if err != nil {
+		return fmt.Errorf("Invalid partition range: %s, %v", latest, err)
+	}
+
+	return db.GeneratePartitionWithTimestamp(ctx, latestTime)
+}
+
+// GeneratePartitionWithTimestamp generates the partition based on the given time
+func (db *DB) GeneratePartitionWithTimestamp(ctx context.Context, begin time.Time) error {
+	monthInterval := defaultPartitionInterval
+	begin = time.Date(begin.Year(), begin.Month(), 1, 0, 0, 0, 0, begin.Location()) // month granularity
+	end := begin.AddDate(0, monthInterval, 0)
+	partitionTableNameSuffix := fmt.Sprintf(`%d_%02dto%d_%02d`, begin.Year(), begin.Month(), end.Year(), end.Month())
+	tableName := fmt.Sprintf(`%s_%s`, tableAWSEventsIPSHostnames, partitionTableNameSuffix)
+
+	tx, err := db.sqldb.BeginTx(ctx, nil)
+	if err != nil {
+		return err
+	}
+	beginFmt := begin.Format(time.RFC3339)
```

Comment:
Question: I was under the impression that the `pg` driver handled `time.Time` types. Is that not correct?

Reply:
Yes it does. But in some places (i.e. partition creation) we can't use prepared statements, rather we must use fmt.Sprintf(). I have updated the prepared statements to use time.Time